### PR TITLE
backup/restore functionality - PVC subset

### PR DESCRIPTION
**Pull request by miguelsorianod**
Merge backup/restore PVC subset new functionality integration branch into master

### CODE REVIEW

File: pkg/apis/apps/v1alpha1/apimanagerbackup_types.go [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+	BackupDestination APIManagerBackupDestination `json:"backupDestination"`
+}
+
+// APIManagerBackupDestination defines the backup data destination
+// configurability. It is a union type. Only one of the fields can be
+// set
+type APIManagerBackupDestination struct {
+	// PersistentVolumeClaim as backup data destination configuration
+	// +optional
+	PersistentVolumeClaim *PersistentVolumeClaimBackupDestination `json:"persistentVolumeClaim,omitempty"`
```
Comment (eguzki) [inline pkg/apis/apps/v1alpha1/apimanagerbackup_types.go:28]: why not actual `PersistentVolumeClaim`

File: pkg/apis/apps/v1alpha1/apimanagerrestore_types.go [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+	RestoreSource APIManagerRestoreSource `json:"restoreSource"`
+}
+
+// APIManagerRestoreSource defines the backup data restore source
+// configurability. It is a union type. Only one of the fields can be
+// set
+type APIManagerRestoreSource struct {
+	// +optional
+	// Restore data soure configuration
+	PersistentVolumeClaim *PersistentVolumeClaimRestoreSource `json:"persistentVolumeClaim,omitempty"`
```
Comment (eguzki) [inline pkg/apis/apps/v1alpha1/apimanagerrestore_types.go:26]: why not actual `PersistentVolumeClaim`

File: doc/operator-backup-and-restore.md [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+     kind: APIManagerBackup
+     metadata:
+      name: example-apimanagerbackup-pvc
+     spec:
+       backupDestination:
+         persistentVolumeClaim:
+           resources:
+             requests: "10Gi"
+   ```
+1. Wait until APIManagerBackup finishes. You can check this by obtaining
```
Comment (eguzki) [inline doc/operator-backup-and-restore.md:70]: `oc wait --for=condition=available --timeout=-1s APIManagerBackup/example-apimanagerbackup-pvc`

File: pkg/apis/apps/v1alpha1/apimanagerbackup_types.go [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+	// Storage Resource requests to be used on the PersistentVolumeClaim.
+	// To learn more about resource requests see:
+	// https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/
+	Requests resource.Quantity `json:"requests"` // Should this be a string or a resoure.Quantity? it seems it is serialized as a string
+}
+
+type APIManagerBackupConditionType string
+
+// APIManagerBackupCondition describes the state of an APIManagerBackup at a certain point.
+type APIManagerBackupCondition struct {
```
Comment (eguzki) [inline pkg/apis/apps/v1alpha1/apimanagerbackup_types.go:64]: I have used this same type for other features:

https://github.com/3scale/3scale-operator/blob/capabilities-v2/pkg/common/status_conditions.go#L57

It does not have anything specific for backup or restore

File: pkg/backup/apimanager_backup_options.go [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+	validator "github.com/go-playground/validator/v10"
+	"k8s.io/apimachinery/pkg/types"
+)
+
+type APIManagerBackupOptions struct {
+	Namespace                  string                      `validate:"required"` // Namespace where the K8s related objects to the backup will be created/looked
+	APIManagerBackupName       string                      `validate:"required"` // Name of the APIManagerBackup CR. NOT the APIManager cr name
+	APIManagerBackupUID        types.UID                   `validate:"required"` // UID of the APIManagerBackup CR
+	APIManagerName             string                      `validate:"required"` // Name of the APIManager CR. NOT the APIManagerBackup cr name
+	APIManager                 *appsv1alpha1.APIManager    `validate:"required"` // Should we make this required?
```
Comment (eguzki) [inline pkg/backup/apimanager_backup_options.go:14]: I think it should be required. APIManager must be available.

File: pkg/backup/apimanager_backup_options_provider.go [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+	if len(resList.Items) > 1 {
+		return nil, fmt.Errorf("Multiple APIManagers found in namespace '%s'. Unsupported scenario", a.APIManagerBackupCR.Namespace)
+	}
+
+	res = &resList.Items[0]
+	return res, nil
+
+}
+
+func (a *APIManagerBackupOptionsProvider) apiManagerFromName(name string) (*appsv1alpha1.APIManager, error) {
```
Comment (eguzki) [inline pkg/backup/apimanager_backup_options_provider.go:99]: called from where??

File: pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go [hunk 7]
```diff
@@ -0,0 +1,534 @@
...
+	if r.cr.Status.StartTime == nil {
+		startTimeUTC := metav1.Time{Time: clock.Now().UTC()}
+		r.cr.Status.StartTime = &startTimeUTC
+		err := r.UpdateResourceStatus(r.cr)
+		return reconcile.Result{Requeue: true}, err
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *APIManagerRestoreLogicReconciler) reconcileRestoreFromS3Source() (reconcile.Result, error) {
```
Comment (eguzki) [inline pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:137]: :shrug:

File: pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go [hunk 8]
```diff
@@ -0,0 +1,534 @@
...
+
+	if errors.IsNotFound(err) {
+		err := r.CreateResource(desired)
+		if err != nil {
+			return reconcile.Result{}, err
+		}
+		return reconcile.Result{Requeue: true, RequeueAfter: 5 * time.Second}, nil
+	}
+
+	// TODO We do not reconcile ownerReference or labels nor annotations
```
Comment (eguzki) [inline pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:215]: Change comment

```
// Jobs ownerReference or labels nor annotations not reconciled
// Jobs are one-shot so there's not much point on making updates to them
```

File: pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go [hunk 9]
```diff
@@ -0,0 +1,534 @@
...
+	secret, err := r.sharedBackupSecret()
+	if err != nil {
+		return nil, err
+	}
+	if secret == nil {
+		return nil, fmt.Errorf("Secret '%s' not found", r.apiManagerRestore.SecretToShareName())
+	}
+
+	// https://godoc.org/k8s.io/apimachinery/pkg/runtime/serializer#CodecFactory
+	codecFactory := k8sserializer.NewCodecFactory(r.Scheme())
```
Comment (eguzki) [inline pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:381]: Why not simply json Marshall/Unmarshall? 

Add a comment explaining why some weird serializer and deserializer is used

File: pkg/restore/apimanager_restore.go [hunk 10]
```diff
@@ -0,0 +1,418 @@
...
+
+const (
+	RestorePVCMountPath           = "/backup"
+	SystemFileStoragePVCMountPath = "/system-filestorage-pvc"
+)
+
+var secretsToRestore map[string]string = map[string]string{
+	"SystemSMTP": "system-smtp",
+	"SystemSeed": "system-seed",
+	// "SystemDatabase":      "system-database", // We decided this is done by the user. Reason for that is that he might want to restore it in another place
```
Comment (eguzki) [inline pkg/restore/apimanager_restore.go:34]: These type of comments are useful for the code?

File: deploy/role.yaml [hunk 11]
```diff
@@ -66,6 +62,12 @@ rules:
   - pods
```
[the remark this answers is not in the document]
Comment (miguelsorianod) [reply deploy/role.yaml:62]: :+1:

File: deploy/role.yaml [hunk 12]
```diff
@@ -77,6 +79,8 @@ rules:
   - apps.3scale.net
   resources:
   - '*'
+  - apimanagerbackups
```
[the remark this answers is not in the document]
Comment (miguelsorianod) [reply deploy/role.yaml:82]: This was automatically added by operator-sdk when creating the CRDs

File: pkg/apis/apps/v1alpha1/apimanagerbackup_types.go (hunk 1 above)
Comment (miguelsorianod) [reply pkg/apis/apps/v1alpha1/apimanagerbackup_types.go:28]: Only a subset of functionality is provided

File: pkg/apis/apps/v1alpha1/apimanagerrestore_types.go (hunk 2 above)
Comment (miguelsorianod) [reply pkg/apis/apps/v1alpha1/apimanagerrestore_types.go:26]: Only a subset of functionality is provided

File: doc/operator-backup-and-restore.md (hunk 3 above)
Comment (miguelsorianod) [reply doc/operator-backup-and-restore.md:70]: This is not a DeploymentConfig.
There's no "available" condition / conditions.

File: pkg/backup/apimanager_backup_options_provider.go (hunk 6 above)
Comment (miguelsorianod) [reply pkg/backup/apimanager_backup_options_provider.go:99]: it was used before but not anymore. I'll delete it :+1:

File: pkg/backup/apimanager_backup_options.go (hunk 5 above)
Comment (miguelsorianod) [reply pkg/backup/apimanager_backup_options.go:14]: :+1: I'll remove the comment

File: pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go (hunk 8 above)
Comment (miguelsorianod) [reply pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:215]: :+1:

File: pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go (hunk 7 above)
Comment (miguelsorianod) [reply pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:137]: :+1:

File: pkg/restore/apimanager_restore.go (hunk 10 above)
Comment (miguelsorianod) [reply pkg/restore/apimanager_restore.go:34]: I think so. It explains why this subset of secrets/configmaps have not been applied

File: pkg/apis/apps/v1alpha1/apimanagerbackup_types.go (hunk 4 above)
Comment (miguelsorianod) [reply pkg/apis/apps/v1alpha1/apimanagerbackup_types.go:64]: :+1:  operator-sdk is also adding that in future versions. If needed that could be used when upgrading

File: pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go (hunk 9 above)
Comment (miguelsorianod) [reply pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:381]: This is just Kubernetes way of deserializing objects into a runtime object (provided by apimachinery). The own decoder.Decode method in apimachinery contains information about it using the kind, version (from scheme) to deserialize to a runtime object

**Comment by miguelsorianod** [comment]
Applied changes and answered comments

File: deploy/role.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (eguzki) [reply deploy/role.yaml:76]: and apimanager is not automatically added? It should be consistent, either all of them or none.

File: doc/operator-backup-and-restore.md (hunk 3 above)
Comment (eguzki) [reply doc/operator-backup-and-restore.md:70]: backup type has condition list. Is not used at all? then shoud be removed?

File: pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go (hunk 9 above)
Comment (eguzki) [reply pkg/controller/apimanagerrestore/apimanagerrestore_logic_reconciler.go:370]: This is comment should be in the code and not lost in github PR

**Comment by eguzki** [comment]
comments left

File: deploy/role.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (miguelsorianod) [reply deploy/role.yaml:76]: :+1: added apimanager and removed wildcard

File: doc/operator-backup-and-restore.md (hunk 3 above)
Comment (miguelsorianod) [reply doc/operator-backup-and-restore.md:70]: Removed

**Comment by miguelsorianod** [comment]
Changes applied

**Comment by eguzki** [comment]
LGTM

Fix failing tests and let's merge

**Comment by miguelsorianod** [comment]
Fixed failing tests: added back the wildcard permission on apps.3scale.net, the removal of it was causing errors creating some objects https://app.circleci.com/pipelines/github/3scale/3scale-operator/2061/workflows/749dbf90-69c4-40d6-821b-2c8d87a2fc2f/jobs/13598

File: deploy/role.yaml (hunk 12 above)
[the remark this answers is not in the document]
Comment (eguzki) [reply deploy/role.yaml]: I do not see apimanagers added

[the remark this answers is not in the document]
Comment (miguelsorianod) [reply deploy/role.yaml]: See my last comment. I did but tests started to fail due to it seems other (sub)resources are needed so I reverted it to * and removed the additional ones to leave it consistent

[the remark this answers is not in the document]
Comment (miguelsorianod) [reply deploy/role.yaml]: Opened Jira issue THREESCALE-2998 to investigate why errors are received when replacing `*` by specific custom resource names

**Pull request merged**